File: test-case/void/186/Revword.kt
import soda.kotlin.unittest.*

class Solution {
    fun reverseWords(s: CharArray): Unit {
        if (s.size == 0) {
            return
        }

        reverse(s, 0, s.size - 1)

        val N = s.size
        var i = 0
        var j = 0
        while (j < N) {
            if (s[j] == ' ') {
                reverse(s, i, j-1)
                i = j + 1
            }
            ++j
        }
        if (i < j) {
            reverse(s, i, j-1)
        }
    }

    private fun reverse(s: CharArray, ii: Int, jj: Int) {
        var i = ii
        var j = jj
        while (i < j) {
            val tmp = s[i]
            s[i] = s[j]
            s[j] = tmp
            ++i
            --j
        }
    }
}

class Revword : (String) -> String {
    override fun invoke(text: String): String {
        // * setup with your method
        // val work = GenericTestWork.create(Solution()::add)

        // * by method that has not return value
        val work = GenericTestWork.createVoid(Solution()::reverseWords)
        // * by class of data struct
        // val work = GenericTestWork.forStruct(STRUCT::class)

        // * setup validator
        // work.validator = (T, T) -> Boolean
        work.compareSerial = true
        return work.run(text)
    }

}

fun main() {
    print(Revword()(Utils.fromStdin()))
}

File: test-case/void/186/Revword.cs
using Soda.Unittest;
using Soda.Leetcode;

public class Solution {
    public void ReverseWords(char[] s)
    {
        if (s.Length == 0) {
            return;
        }

        reverse(s, 0, s.Length - 1);

        int N = s.Length;
        int i = 0, j = 0;
        while (j < N) {
            if (s[j] == ' ') {
                reverse(s, i, j-1);
                i = j + 1;
            }
            ++j;
        }
        if (i < j) {
            reverse(s, i, j-1);
        }
    }

    private void reverse(char[] s, int i, int j) {
        while (i < j) {
            (s[i], s[j]) = (s[j], s[i]);
            ++i;
            --j;
        }
    }
}

public class Revword
{
    public static void Main(string[] args)
    {
        var work = WorkFactory.Create(Utils.Fn(new Solution().ReverseWords));
        // var work = WorkFactory.ForStruct<STRUCT>();
        // work.SetValidator((e, r) => ...);
        work.CompareSerial = true;
        Console.WriteLine(work.Run(Utils.ReadStdin()));
    }
}
